chaos/chaos.go

```diff
@@ -48,104 +47,94 @@ func (c *Chaos) DurationToKillTime() time.Duration {
 	return c.killAt.Sub(time.Now())
 }
 
-// Does the actual execution of the chaos, i.e.
-// termination of pods
+// Exposed function that calls the actual execution of the chaos, i.e. termination of pods
 // The result is sent back over the channel provided
 func (c *Chaos) Execute(resultchan chan<- *ChaosResult) {
-	// Create kubernetes client
-	client, err := CreateClient()
+	// Create kubernetes clientset
+	clientset, err := kubernetes.CreateClient()
 	if err != nil {
 		resultchan <- c.NewResult(err)
 		return
 	}
 
-	// Is deployment still enrolled in kube-monkey
-	enrolled, err := c.deployment.IsEnrolled(client)
+	err = c.verifyExecution(clientset)
 	if err != nil {
 		resultchan <- c.NewResult(err)
 		return
 	}
-	if !enrolled {
-		resultchan <- c.NewResult(fmt.Errorf("Skipping deployment %s; reason=no longer enrolled", c.deployment.Name()))
-		return
-	}
-
-	// Has deployment been blacklisted since scheduling?
-	if c.deployment.IsBlacklisted(config.BlacklistedNamespaces()) {
-		resultchan <- c.NewResult(fmt.Errorf("Skipping deployment %s; reason=blacklisted", c.deployment.Name()))
-		return
-	}
-
-	// Do the termination
-	killAll, err := c.deployment.HasKillAll(client)
-	if err != nil {
-		glog.Errorf("Failed to check KillAll label for deployment %s. Proceeding with termination of a single pod. Error: %v", c.deployment.Name(), err.Error())
-	}
-
-	if killAll {
-		err = c.TerminateAll(client)
-	} else {
-		err = c.Terminate(client)
-	}
 
+	err = c.terminate(clientset)
 	if err != nil {
 		resultchan <- c.NewResult(err)
-	} else {
-		// Send a success msg
-		resultchan <- c.NewResult(nil)
+		return
 	}
+
+	// Send a success msg
+	resultchan <- c.NewResult(nil)
 }
 
-// Runs the actual pod-termination logic
-func (c *Chaos) Terminate(client *kube.Clientset) error {
-	// Pick a target pod to delete
-	pods, err := c.deployment.RunningPods(client)
+// Verify if the victim has opted out since scheduling
+func (c *Chaos) verifyExecution(clientset *kube.Clientset) error {
+	// Is victim still enrolled in kube-monkey
+	enrolled, err := c.Victim().IsEnrolled(clientset)
 	if err != nil {
 		return err
 	}
 
-	if len(pods) == 0 {
-		return fmt.Errorf("Deployment %s has no running pods at the moment", c.deployment.Name())
+	if !enrolled {
+		return fmt.Errorf("%s %s is no longer enrolled in kube-monkey. Skipping\n", c.Victim().Kind(), c.Victim().Name())
 	}
 
-	targetPod := RandomPodName(pods)
+	// Has the victim been blacklisted since scheduling?
+	if c.Victim().IsBlacklisted() {
+		return fmt.Errorf("%s %s is blacklisted. Skipping\n", c.Victim().Kind(), c.Victim().Name())
+	}
 
-	glog.V(2).Infof("Terminating pod %s for deployment %s", targetPod, c.deployment.Name())
-	return c.DeletePod(client, targetPod)
-}
+	// Has the victim been removed from the whitelist since scheduling?
+	if !c.Victim().IsWhitelisted() {
```

Comment:
@Spellchaser  Does this mean that each victim has to specifically be listed in the whitelist for kube-monkey to act on it?

Reply:
Yes, but the default for whitelisting is all namespaces. So if you don't enable a whitelist, it doesn't affect you.

Reply:
I suppose a better comment would be something like
`//  Does the namespace whitelist still include the victim's namespace since scheduling? `

Reply:
It looks like the default is `NamespaceDefault`

```
viper.SetDefault(param.WhitelistedNamespaces, []string{metav1.NamespaceDefault})
```

Is that the same as `NamespaceAll`?

Reply:
Nada, NamespaceDefault is "default" NamespaceAll is ""
I figured it'd be safer, abliet more confusing for users

Reply:
So, if I don't specify the `whitelisted_namespaces` in my config file, my victims will not be killed (assuming they are in their own namespace) ?

Reply:
Ok, ignore that. I saw the following commit where you update docs for the whitelisting stuff and that clears it up.